overlord/devicestate/devicemgr.go

```diff
@@ -0,0 +1,384 @@
+	resp, err := cli.Post(serialRequestURL, asserts.MediaType, bytes.NewBuffer(encodedSerialRequest))
+	if err != nil {
+		return nil, &state.Retry{After: 60 * time.Second}
+	}
+	if resp.StatusCode >= 400 && resp.StatusCode < 500 {
+		return nil, fmt.Errorf("unexpected status code %d trying to post serial request to %s)", resp.StatusCode, serialRequestURL)
+	}
+
+	switch resp.StatusCode {
+	case 200, 201:
+	case 202:
+		return nil, errPoll
+	default:
+		return nil, &state.Retry{After: 60 * time.Second}
+	}
+
+	// decode body with serial assertion
+	dec := asserts.NewDecoder(resp.Body)
+	got, err := dec.Decode()
+	if err != nil { // assume broken i/o
+		return nil, &state.Retry{After: 60 * time.Second}
+	}
+
+	serial, ok := got.(*asserts.Serial)
+	if !ok {
+		return nil, fmt.Errorf("serial-request response did not return a serial assertion")
+	}
+
+	return serial, nil
+}
+
+func doDeliverSerialRequest(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	devIdentity, err := auth.Device(st)
+	if err != nil {
+		return err
+	}
+
+	var retrieveSerialID string
+	err = t.Get("serial-setup-task", &retrieveSerialID)
+	if err != nil {
+		return err
+	}
+	retrieveSerial := st.Task(retrieveSerialID)
+	if retrieveSerial == nil {
+		return fmt.Errorf("internal error: cannot find retrieve serial task")
+	}
+
+	pk, err := keyPair(st)
+	if err != nil {
+		return err
+	}
+
+	cli := &http.Client{Timeout: 30 * time.Second}
+
+	st.Unlock()
+	resp, err := cli.Get(serialRequestURL)
+	st.Lock()
+	if err != nil {
+		return &state.Retry{After: 60 * time.Second}
+	}
+	if resp.StatusCode >= 400 && resp.StatusCode < 500 {
+		return fmt.Errorf("unexpected status code %d trying to get serial request nonce/ticket (from %s)", resp.StatusCode, serialRequestURL)
+	}
+	if resp.StatusCode != 200 {
+		return &state.Retry{After: 60 * time.Second}
+	}
+
+	dec := json.NewDecoder(resp.Body)
+	var nonceTicket nonceTicketResp
+	err = dec.Decode(&nonceTicket)
+	if err != nil { // assume broken i/o
+		return &state.Retry{After: 60 * time.Second}
+	}
+
+	// XXX: do it the hard way for now (should be much simpler soon)
+	privKey := asserts.OpenPGPPrivateKey(packet.NewRSAPrivateKey(time.Unix(1, 0), pk))
+	encodedPubKey, err := asserts.EncodePublicKey(privKey.PublicKey())
+	if err != nil {
+		return fmt.Errorf("internal error: cannot encode device public key: %v", err)
+
+	}
+
+	serialReq, err := asserts.FreestandingSign(asserts.SerialRequestType, map[string]interface{}{
+		"brand-id":     devIdentity.Brand,
+		"model":        devIdentity.Model,
+		"nonce-ticket": nonceTicket.NonceTicket,
+		"device-key":   string(encodedPubKey),
+	}, nil, privKey) // XXX: fill body with some agreed hardware details
+	if err != nil {
+		return err
+	}
+
+	encodedSerialReq := asserts.Encode(serialReq)
+
+	// NB: until we get at least an Accepted (202) we need to
+	// retry from scratch creating a new nonce because the
+	// previous one used could have expired
```

Comment:
By now it feels like there's a lot of back and forth between these tasks with pretty much no gain. Have you considered having a single task that submits and installs the serial at once, and asks for a retrial when necessary? It would probably avoid a lot of the boilerplate we see here, probably without disadvantages.